packages/vm/lib/buildBlock.ts

```diff
@@ -0,0 +1,241 @@
+import { Address, BN, toBuffer } from 'ethereumjs-util'
+import { encode } from 'rlp'
+import { BaseTrie as Trie } from 'merkle-patricia-tree'
+import { TypedTransaction } from '@ethereumjs/tx'
+import { Block, BlockOptions, HeaderData } from '@ethereumjs/block'
+import VM from '.'
+import { RunTxResult } from './runTx'
+import Bloom from './bloom'
+import { generateTxReceipt, calculateMinerReward, rewardAccount } from './runBlock'
+
+/**
+ * Options for building a block.
+ */
+export interface BuildBlockOpts {
+  /**
+   * The parent block
+   */
+  parentBlock: Block
+
+  /**
+   * The block header data to use.
+   * Defaults used for any values not provided.
+   */
+  headerData?: HeaderData
+
+  /**
+   * The block options to use.
+   */
+  blockOpts?: BlockOptions
+}
+
+/**
+ * Options for sealing a block.
+ */
+export interface SealBlockOpts {
+  /**
+   * For PoW, the nonce.
+   * Overrides the value passed in the constructor.
+   */
+  nonce?: Buffer
+
+  /**
+   * For PoW, the mixHash.
+   * Overrides the value passed in the constructor.
+   */
+  mixHash?: Buffer
+}
+
+export class BlockBuilder {
+  private readonly vm: VM
+  private blockOpts: BlockOptions
+  private headerData: HeaderData
+  private transactions: TypedTransaction[] = []
+  private transactionResults: RunTxResult[] = []
+  private checkpointed = false
+  private reverted = false
+  private built = false
+
+  constructor(vm: VM, opts: BuildBlockOpts) {
+    this.vm = vm
+    this.blockOpts = { ...opts.blockOpts, common: this.vm._common }
+
+    this.headerData = {
+      ...opts.headerData,
+      parentHash: opts.headerData?.parentHash ?? opts.parentBlock.hash(),
+      number: opts.headerData?.number ?? opts.parentBlock.header.number.addn(1),
+      gasLimit: opts.headerData?.gasLimit ?? opts.parentBlock.header.gasLimit,
+    }
+  }
+
+  /**
+   * Throws if the block has already been built or reverted.
+   */
+  checkStatus() {
+    if (this.built) {
+      throw new Error('Block has already been built')
+    }
+    if (this.reverted) {
+      throw new Error('State has already been reverted')
+    }
+  }
+
+  /**
+   * Returns the cumulative gas used by the transactions added to the block.
+   */
+  gasUsed(): BN {
+    return this.transactionResults.reduce(
```

Comment:
I'd accumulate this explicitly as txs are executed, instead of computing it on demand. Otherwise, when you go adding txs, you end up with a quadratic operation.

Reply:
The reason for this is that to select a txs we need to call this method, as we can only add txs that fit in `blockGasLimit - builder.gasUsed()`.

Reply:
good idea, updated in https://github.com/ethereumjs/ethereumjs-monorepo/pull/1158/commits/eb197ab68b74a61a52b1c61931599fc9610a5b40